File: app/src/main/java/com/sp/base/ui/model/MainViewModel.java
package com.sp.base.ui.model;


import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;


import com.sp.base.network.ApiResponse;
import com.sp.base.network.Repository;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.schedulers.Schedulers;


public class MainViewModel extends ViewModel {
    public Repository repository;
    private final MutableLiveData<ApiResponse> responseLiveData = new MutableLiveData<>();

    @ViewModelInject
    public MainViewModel(Repository repository) {
        this.repository = repository;
    }

    public MutableLiveData<ApiResponse> getResponse() {
        return responseLiveData;
    }

//    @SuppressLint("CheckResult")
//    public void getRouteList(GetRouteListRequest getRouteListRequest) {
//        repository.executeGetRouteListAPI(getRouteListRequest)
//                .subscribeOn(Schedulers.io())
//                .observeOn(AndroidSchedulers.mainThread())
//                .doOnSubscribe((d) -> responseLiveData.setValue(ApiResponse.loading()))
//                .subscribe(
//                        result -> responseLiveData.setValue(ApiResponse.success(result)),
//                        throwable -> responseLiveData.setValue(ApiResponse.error(throwable))
//                );
//
//    }

    public void getDataList() {
        repository.executeGetDataList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe((d) -> responseLiveData.setValue(ApiResponse.loading()))
                .subscribe(
                        result -> responseLiveData.setValue(ApiResponse.success(result)),
                        throwable -> responseLiveData.setValue(ApiResponse.error(throwable))
                );
    }

}

File: app/src/main/java/com/baseapplication/ui/model/MainViewModel.kt
package com.onetuchservice.business.ui.model

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.paging.PagedList
import com.baseapplication.network.ApiResponse
import com.baseapplication.network.Repository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class MainViewModel @Inject constructor(var repository: Repository) : ViewModel() {
    private val TAG = MainViewModel::class.java.simpleName

    // PagedList configuration
    private val config = PagedList.Config.Builder()
        .setPageSize(20) // Page size (the number of items to load per page)
        .setEnablePlaceholders(false)
        .build()
    // LiveData to observe the API result
    private val _responseLiveData = MutableLiveData<ApiResponse?>()
    val responseLiveData get() = _responseLiveData

    //using kotlin coroutins
    fun getData(page: Int, limit: Int) {
       viewModelScope.launch {
           val response =   repository.callGetDataList(page, limit)
           _responseLiveData.value = response
       }
    }

    //using RxJava
        fun getDataList(page: Int, limit: Int) {
        repository.executeGetDymmyList(page, limit)?.doOnSubscribe {
            responseLiveData.setValue(ApiResponse.loading())
        }?.subscribe({ result ->
            // On Success
            responseLiveData.setValue(ApiResponse.success(result))
        }, { throwable ->
            // On Error
            responseLiveData.setValue(ApiResponse.error(throwable))
        })
    }

}


